var/spack/repos/builtin/packages/python/package.py

```diff
@@ -1100,20 +1106,33 @@ def libs(self):
 
     @property
     def headers(self):
-        # Location where pyconfig.h is _supposed_ to be
-        config_h = self.config_vars["config_h_filename"]
-        if os.path.exists(config_h):
-            headers = HeaderList(config_h)
+        # Locations where pyconfig.h could be
+        # This varies by system, especially on macOS where the command line tools are
+        # installed in a very different directory from the system python interpreter.
+        py_version = str(self.version.up_to(2))
+        candidates = list(
+            dedupe(
+                [
+                    os.path.dirname(self.config_vars["config_h_filename"]),
+                    self.config_vars["INCLUDEPY"],
+                    self.config_vars["CONFINCLUDEPY"],
+                    os.path.join(self.config_vars["base"], "include", py_version),
+                    os.path.join(self.config_vars["base"], "Headers"),
+                ]
+            )
+        )
+
+        for directory in candidates:
+            headers = find_headers("pyconfig", directory)
+            if headers:
+                config_h = headers[0]
+                break
         else:
-            # If not, one of these config vars should contain the right directory
-            for var in ["INCLUDEPY", "CONFINCLUDEPY"]:
-                headers = find_headers("pyconfig", self.config_vars[var])
-                if headers:
-                    config_h = headers[0]
-                    break
-            else:
-                msg = "Unable to locate {} headers in {}"
-                raise spack.error.NoHeadersError(msg.format(self.name, directory))
+            raise spack.error.NoHeadersError(
+                "Unable to locate {} headers in {}".format(
+                    self.name, self.config_vars["INCLUDEPY"]
+                )
+            )
 
         headers.directories = [os.path.dirname(config_h)]
```

Comment:
Is this correct for all layouts? This was intended for layouts like:
```
<prefix>/include/pythonX.Y/pyconfig.h
```
where a user would use:
```c
#include <pythonX.Y/pyconfig>
```
But if the location is something like:
```
<prefix>/Headers/pyconfig.h
```
what would the proper directory be?

Reply:
It doesn't really matter as long as you find the python includes.  On macOS, `include/python3.8` is a symlink to `Headers` in the framework directory, so we would find the `include/python3.8` option first, given the order I listed things.

I'm not sure what the "proper" directory is, but I prefer to find the ones that looks more like a standard unix layout first, because it's more likely that code that uses `..` to infer things from discovered paths will be correct.  e.g., I think most code in a generic build assumes that python headers will be in `include/pythonX.Y` and that `bin`, `lib`, etc. are in `../..` from there.  If you return the `Headers` directory, that breaks because it's only one deep (because Apple or something -- I guess it's b/c that is how their frameworks work -- you pick the python you want, THEN include its `Headers` directory.

Reply:
Maybe we should change this to:
```suggestion
        if not headers.directories[0].endswith('Headers'):
            headers.directories = [os.path.dirname(config_h)]
```

Reply:
oh I see what you're saying -- nice catch.  Yeah lemme see if I can either do that or simplify it further.

Reply:
Nvm, this is correct as is.